File: time independence/matrix_free.py
"""
Code for diagonalizing a Hamiltonian of the form H = P^2 + V(x).
The code uses the power method using the conjugate gradient as a method for inverting a matrix.
Everything is implemented in the matrixless form, i.e. no matrices are allocated
instead whenever necessary the result of the matrix-vector product is calculated.
"""
import time
import numpy as np
from math import factorial
import matplotlib.pyplot as plt
from scipy.special import eval_hermite


def conj_grad_matrixless(mult, b, tol=1e-6):
    """
    Matrix-free implementation of conjugate gradient method for solving M v = b.
    
    Parameters
    ----------
    mult : callable
        A function that returns the product of the matrix with a vector (matrix-free approach).
    b : 1darray
        Ordinate or "dependent variable" values.
    tol : float, optional
        Required tolerance (default 1e-6).

    Return
    ------
    x : 1darray
        Solution of the system.
    err : float
        Error of the solution.
    """

    N = len(b)
    x = np.zeros(N)  # Initial guess

    r  = b - mult(x) # Residuals
    p  = r           # Descent direction
    r2 = sum(r*r)    # Norm^2 of residuals

    iter = 0

    while True:
        Ap = mult(p)              # Compute matrix-free product M @ p
        alpha = r2 / (p @ Ap)     # Descent step

        x = x + alpha * p         # Update position
        r = r - alpha * Ap        # Update residuals

        r2_new = sum(r*r)         # Norm^2 of new residuals
        beta = r2_new / r2        # Compute step for p

        r2 = r2_new  # Update norm

        if np.sqrt(r2_new) < tol:  # Break condition
            break

        p = r + beta * p  # Update descent direction
        iter += 1

    
    err = np.sqrt(r2_new)
    return x, err


def eig_matrixless(mult, N, k=None, tol=1e-10, magnitude='small', inverse_solver='cg'):
    '''
    Compute the eigenvalue decomposition of a matrix using power iteration
    or inverse iteration in a matrix-free manner, avoiding storing the matrix.

    Parameters
    ----------
    mult : Callable
        A function that returns the product of the matrix with a vector (matrix-free approach).
    k : int or None
        Number of eigenvalues and eigenvectors to find. If None, find all.
    tol : float
        Tolerance for convergence.
    magnitude : string
        'small' to find smallest eigenvalues, 'big' to find largest eigenvalues.
    inverse_solver : string
        Method for solving linear systems. Default is 'cg' (Conjugate Gradient).
    
    Returns
    -------
    eigval : 1darray
        Eigenvalues found.
    eigvec : 2darray
        Corresponding eigenvectors.
    counts : 1darray
        Number of iterations for each eigenvalue.
    '''

    # Handle whether we want the smallest or largest eigenvalues
    if magnitude == 'small':
        # Define a matrix-free inverse operator using your Conjugate Gradient implementation
        def mult_inv(v):
            if inverse_solver == 'cg':
                # Use custom CG solver to solve M*x = v (instead of np.linalg.solve)
                x, _ = conj_grad_matrixless(mult, v)
                return x
            else:
                raise NotImplementedError(f"Inverse solver '{inverse_solver}' not implemented")
        
        mat = mult_inv
    else:
        # Normal power iteration
        mat = mult

    if k is None:
        k = N

    eigvec = []
    eigval = []
    counts = []

    for _ in range(k):

        # Initializzation
        v_p = np.random.randn(N)
        v_p = v_p / np.linalg.norm(v_p)
        l_v = np.random.random()
        Iter = 0

        while True:
            l_o = l_v
            v_o = v_p                   # Update vector
            v_p = mat(v_p)              # Compute new vector
            v_p /= np.linalg.norm(v_p)  # Normalizzation
            
            # Orthogonalization respect
            # all eigenvectors find previously
            for i in range(len(eigvec)):
                v_p = v_p - np.dot(eigvec[i], v_p) * eigvec[i]

            # Eigenvalue of v_p, A @ v_p = l_v * v_p
            # Multiplying by the transposed => (A @ v_p) @ v_p.T = l_v
            # Using v_p @ v_p.T = 1
            l_v = np.dot(v_p, mat(v_p))
            
            R1 = np.linalg.norm(v_p - v_o)
            R2 = np.linalg.norm(v_o + v_p)
            R3 = abs(l_v - l_o)               # In eigenvalues the convergence is quadratic
            
            Iter += 1
            if R1 < tol or R2 < tol or R3 < tol:
                break

        eigvec.append(v_p)
        eigval.append(l_v)
        counts.append(Iter)

    if magnitude == 'small':
        eigval = 1 / np.array(eigval)
    else:
        eigval = np.array(eigval)

    eigvec = np.array(eigvec).T
    return eigval, eigvec, counts



if __name__ == "__main__":

    np.random.seed(69420)
    
#===============================================================================
# Computational parameter
#===============================================================================
         
    k  = 10                    # how many levels compute
    n  = 1000                  # size of matrix
    xr = 10                    # bound
    xl = -10                   # bound
    L  = xr - xl               # dimension of box
    h  = (xr - xl)/(n)         # step size
    tt = np.linspace(0, n, n)  # array form 0 to n
    xp = xl + h*tt             # array of position
    
#===============================================================================
# Hamiltonian
#===============================================================================

    def H(psi, V, x):
        '''
        This function computes the product M @ v without storing M.
        In this case we have a simple tridiagonal matrix of the form:
        
                  [ 2 -1  0  0  0 ]       [ V0 0  0  0  0 ]
                  [-1  2 -1  0  0 ]       [ 0  V1 0  0  0 ]
        1/(2*h^2) [ 0 -1  2 -1  0 ]   +   [ 0  0  V3 0  0 ]
                  [ 0  0 -1  2 -1 ]       [ 0  0  0  V4 0 ]
                  [ 0  0  0 -1  2 ]       [ 0  0  0  0  V5]
        
        Parameters
        ----------
        psi : 1darray
            vector to multiply to the matrix
        V : callable
            potential of the sistem
        x : 1darray
            grid of our discretized sistem
        
        Return
        ------
        H_psi : 1darray
            result of H @ psi
        '''
        N = len(x)
        h = np.diff(x)[0]
        H_psi = np.zeros(N)
        
        # Kinetic term (tridiagonal part)
        factor = - 1/ (2 * h**2)
        
        for i in range(1, N-1):  # Loop through interior points only
            H_psi[i] = psi[i - 1] - 2 * psi[i] + psi[i + 1]
        
        # Boundary conditions
        H_psi[0]   =  -2 * psi[0]   + psi[1]
        H_psi[N-1] =  -2 * psi[N-1] + psi[N-2]
        
        H_psi *= factor

        # Potential term (diagonal part)
        H_psi += V(x) * psi
        
        return H_psi
    
    def Potential(x):
        return 0.5 * x**2

#===============================================================================
# Computation
#===============================================================================
    
    start = time.time()
    eigval, eigvec, Iter = eig_matrixless(lambda v: H(v, Potential, xp), n, k, tol=1e-5, magnitude='small')

    end = time.time() - start
    print(f'Elapsed time       = {end:.5f}\n')
    
    print("Theoretical     Computed          error")
    print("-----------------------------------------")
    for i in range(k):
        print(f'{i+0.5} \t \t {eigval[i]:.5f} \t {eigval[i]-(i+0.5):.2e}')
    
    psi = eigvec/np.sqrt(h)
    
#===============================================================================
# Plot
#===============================================================================
    
    def G(x, m):
        return (1/(np.pi)**(1/4))*(1/np.sqrt((2**m)*factorial(m)))*eval_hermite(m, x)*np.exp(-(x**2)/2)
   
    plt.figure(1)
    plt.title("Oscillatore armonico")#, fontsize=15)
    plt.xlabel('x')#, fontsize=15)
    plt.ylabel(r'$|\psi(x)|^2$')#, fontsize=15)
    plt.grid()
    plt.ylim(0, 3)
    plt.xlim(-5, 5)

    plt.plot(xp, Potential(xp) , color='black', label='V(x)= $ x^2/2 $')
    c = ['b', 'r', 'g']
    
    for L in range(3):
        
        plt.errorbar(xp, abs(psi[:,L])**2 + eigval[L], color=c[L], fmt='.')
        plt.plot(xp, np.ones(len(xp))*eigval[L], color=c[L], linestyle='--', label='$E_{%d}=%f$' %(L, eigval[L]))
        plt.plot(xp, abs(G(xp, L))**2 + eigval[L], color='k')
        
    #plt.legend(loc='best')
    
    plt.figure(2)
    plt.title("$\psi(x)$", fontsize=20)
    plt.xlabel('x', fontsize=15)
    plt.ylabel('$\psi(x)$', fontsize=15)
    plt.grid()

    c = ['b', 'r', 'g']
    
    for L in range(3):
        plt.errorbar(xp, abs(abs(psi[:,L])**2 - abs(G(xp, L))**2), color=c[L], fmt='.')

    plt.show()

File: Schr_eq/matrix_free.f90
program matrix_free_diagnalizzation

    ! Code for diagonalizing a Hamiltonian of the form H = P^2 + V(x).
    ! The code uses the power method using the conjugate gradient as a method for inverting a matrix.
    ! Everything is implemented in the matrixless form, i.e. no matrices are allocated
    ! instead whenever necessary the result of the matrix-vector product is calculated.
    ! This code is the translation of matrix_free.py

    real*8 , dimension(:), allocatable :: xp            ! Grid point
    real*8 , dimension(:), allocatable :: eigval        ! To store eigenvalues
    real*8 , dimension(:, :), allocatable :: eigvec     ! To store eigenvectors
    real*8 :: tol, h, xl, xr
    common h

    open(unit=10, file='mf_val.dat', status="unknown")  ! File for eigenvalues
    open(unit=20, file='mf_vec.dat', status="unknown")  ! File for eigenvectors

    call cpu_time(start)

    N   = 10000                               ! Number of grid points
    k   = 5                                   ! Number of eigen(values/vectors) to compute
    xl  = -10.d0                              ! Lower limit on x axis
    xr  = 10.d0                               ! Upper limit on x axis
    h   = (xr - xl) / ((N-1) * 1.d0)          ! Grid spacing
    tol = 1.d-10

    allocate(xp(N), eigval(N), eigvec(N, k))

    do i = 1, N
        xp(i) = xl + h * ((i-1)*1.d0)         ! Array of positions
    enddo

    call eig_matrixless(N, k, tol, eigval, eigvec, xp)

    ! Save on file
    print '("      Theoretical                 Computed                      error")'
    do i = 1, k
        write(10, *) eigval(i)
        write(20, *) eigvec(:, i)/dsqrt(h)
        print *, (i-1) + 0.5d0, eigval(i), eigval(i) - ((i-1) + 0.5d0)
    enddo

    write(20, *) xp

    deallocate(xp, eigval, eigvec)
    close(10)
    close(20)

    call cpu_time(finish)
    print '("Elapsed time = ", f16.8," secondi.")', finish-start

end program


subroutine conj_grad_matrixless(N, b, x, xp, tol)

    !===============================================================================
    ! Matrix-free implementation of conjugate gradient method for solving M x = b.
    ! The function that implements matrix vector multiplication
    ! is the H_psi subroutine defined below.
    !
    ! Parameters
    ! ----------
    ! N : int 
    !     size of the input and output vector
    ! b : 1darray
    !     Ordinate or "dependent variable" values.
    ! x : 1darray
    !     Solution of the sistem, i.e. the output of the subroutine
    ! xp : 1darray
    !     grid of our discretized sistem
    ! tol : float, optional
    !     Required tolerance
    !===============================================================================
    
    real*8, dimension(N) :: b, x, xp, r, p, Hpsi
    real*8 alpha, beta, r2, r2_new, tol

    call H_psi(N, x, r, xp)
    
    r  = b - r                ! Residual
    p  = r                    ! Descent direction
    r2 = dot_product(r, r)    ! Norm^2 of residuals

    do while (dsqrt(r2) > tol)
        
        call H_psi(N, p, Hpsi, xp)         ! Compute matrix-free product M @ p
        
        alpha = r2 / dot_product(p, Hpsi)  ! Descent step
        
        x = x + alpha * p                  ! Update position
        r = r - alpha * Hpsi               ! Update residuals
        
        r2_new = dot_product(r, r)         ! Norm^2 of new residuals
        beta   = r2_new / r2               ! Compute step for p

        p  = r + beta * p                  ! Update descent direction
        r2 = r2_new                        ! Update residuals norm
       
    enddo
    
    return
end

subroutine eig_matrixless(N, k, tol, eigval, eigvec, xp)

    !===============================================================================
    ! Compute the eigenvalue decomposition of a matrix using inverse iteration 
    ! in a matrix-free manner, avoiding storing the matrix.
    !
    ! Parameters
    ! ----------
    ! N : int 
    !     size of the input and output vector
    ! k : int
    !     Number of eigen(values/vectors) to compute
    ! tol : float, optional
    !     Required tolerance
    ! eigval : 1darray
    !     eigenvalues found, stored in an array
    ! eigvec : 2darray
    !     Corresponding eigenvectors matrix
    ! xp : 1darray
    !     grid of our discretized sistem
    !===============================================================================

    real*8 , dimension(N) :: xp
    real*8 , dimension(N) :: eigval
    real*8 , dimension(N, k) :: eigvec
    real*8 , dimension(N) :: v_p, v_o, v_l
    real*8 :: tol, l_v, l_o, R1, R2, R3, R

    character :: cr  ! Variable for loading
    cr = char(13)    ! is used to override the shell
    
    do i = 1, k

        write(*, '(A1)', advance='no')  cr   ! I write cr needed to clean the shell

        R1 = 1.d0
        R2 = 1.d0
        R3 = 1.d0
        R  = 1.d0

        ! initializzation
        call random_number(v_p)
        v_p = v_p / dsqrt(dot_product(v_p, v_p))
        l_v = 0.d0

        do while ((R > tol).or.(R3 > tol))
            
            l_o = l_v                                       ! Update
            v_o = v_p
            
            call conj_grad_matrixless(N, v_p, v_p, xp, tol) ! Compute new vector

            v_p = v_p / dsqrt(dot_product(v_p, v_p))

            ! Orthogonalization respect all eigenvectors find previously
            do j = 1, i - 1
                v_p = v_p - dot_product(eigvec(:, j), v_p) * eigvec(:, j)
            end do
            v_p = v_p / sqrt(dot_product(v_p, v_p))

            ! Eigenvalue of v_p: A @ v_p = l_v * v_p
            ! Multiplying by the transposed => (A @ v_p) @ v_p.T = l_v
            ! Using v_p @ v_p.T = 1
            call conj_grad_matrixless(N, v_p, v_l, xp, tol)
            l_v = dot_product(v_p, v_l)

            R1 = dsqrt(dot_product(v_p - v_o, v_p - v_o))
            R2 = dsqrt(dot_product(v_o + v_p, v_o + v_p))
            R3 = dabs(l_v - l_o)  !In eigenvalues the convergence is quadratic
            R  = dmin1(R1, R2)
            
        enddo

        eigvec(:, i) = v_p
        eigval(i) = 1.d0 / l_v

        ! Loading section, I write percentage
        write(*,'(f8.1, a2)', advance='NO') i/float(k)*100," %"
        flush(6) ! Clean the line
        
    enddo

    write(*,*) ! To avoid problem in printing

    return
end

subroutine H_psi(N, psi, Hpsi, xp)

    !===============================================================================
    ! This function computes the product M @ v without storing M.
    ! In this case we have a simple tridiagonal matrix of the form:
    !    
    !           [ 2 -1  0  0  0 ]       [ V0 0  0  0  0 ]
    !           [-1  2 -1  0  0 ]       [ 0  V1 0  0  0 ]
    ! 1/(2*h^2) [ 0 -1  2 -1  0 ]   +   [ 0  0  V3 0  0 ]
    !           [ 0  0 -1  2 -1 ]       [ 0  0  0  V4 0 ]
    !           [ 0  0  0 -1  2 ]       [ 0  0  0  0  V5]
    !    
    ! Parameters
    ! ----------
    ! psi : 1darray
    !     vector to multiply to the matrix
    ! Hpsi : 1darray
    !     Result of the multiplication
    ! xp : 1darray
    !     grid of our discretized sistem
    !    
    !===============================================================================

    real*8 , dimension(N) :: xp, psi, Hpsi
    real*8 :: fact, V, h
    common h

    !**************** Kinetic part ****************
    fact = -0.5d0 / h**2

    Hpsi(1) = fact * (-2.0d0 * psi(1) + psi(2))
    do i = 2, N - 1
        Hpsi(i) = fact * (psi(i - 1) - 2.0d0 * psi(i) + psi(i + 1))
    enddo
    Hpsi(N) = fact * (-2.0d0 * psi(N) + psi(N - 1))

    !**************** Potential part ****************
    do i = 1, N
        V = 0.5d0 * xp(i)**2
        Hpsi(i) = Hpsi(i) + V * psi(i)
    enddo

    return
end
